### PR TITLE
feature: Implement CompactDifficulty to Work

### PR DESCRIPTION
**Pull request by teor2345**
This is the first step in calculating the cumulative work for each chain.
We use the cumulative work to pick the best chain.

### CODE REVIEW

File: zebra-chain/src/block/difficulty.rs
Comment (yaahc) [inline zebra-chain/src/block/difficulty.rs:110]: :+1: I think we have a while before we gotta fix this one... lol

File: zebra-chain/src/block/difficulty.rs
Comment (yaahc) [inline zebra-chain/src/block/difficulty.rs]: This seems very strange, I don't know of a better way to handle this but I feel like there has to be some way to do log2 with a u128. Gonna look at some math crates. This isn't a blocker, I don't think it matters how nice the impl for this is given that its just for a `Debug` impl.

File: zebra-chain/src/block/difficulty.rs
Comment (teor2345) [reply zebra-chain/src/block/difficulty.rs:110]: If my rough sums are correct, it would take over a million times earth's current energy usage to do this much work. Or about a thousand times the energy earth gets from the sun.

So yeah. 2^64 is a bit risky, 2^128 is way too big.

File: zebra-chain/src/block/difficulty.rs
Comment (teor2345) [reply zebra-chain/src/block/difficulty.rs]: Yeah I was surprised the standard library doesn't have u128 to f64. (Or even lossy u64 to f64, which would work for now.)

But I don't think it's worth adding a dependency for a debug impl.

**Review by yaahc — COMMENTED** [review]
this LGTM but I don't feel super confident reviewing crypto adjacent code so I'm going to defer to @hdevalence or @dconnolly for the actual approval.

File: zebra-chain/src/block/difficulty.rs
Comment (m-ou-se) [reply zebra-chain/src/block/difficulty.rs]: The standard library doesn't expose it, but the language does (through `as`):
```rust
(self.0 as f64).log2()
```

Comment (yaahc) [inline zebra-chain/src/block/difficulty.rs:116]: I've been informed that this also "just works"
```suggestion
        let log2_work = (self.0 as f64).log2();
```

Comment (teor2345) [reply zebra-chain/src/block/difficulty.rs]: Thanks! I'll go fix that up.

I looked at f64:
https://doc.rust-lang.org/std/primitive.f64.html

But not at the type cast reference:
https://doc.rust-lang.org/reference/expressions/operator-expr.html#type-cast-expressions

Comment (yaahc) [reply zebra-chain/src/block/difficulty.rs]: > * Enum cast
>     * Casts an enum to its discriminant, then uses a numeric cast if needed.

:eyes: ooo, what's this?

```rust
enum Foo {
    Baz,
    Bar,
}

fn main() {
    let e = Foo::Bar;
    let d = e as u8;
    println!("{:?}", d);
}
```

TIL this works

Comment (teor2345) [reply zebra-chain/src/block/difficulty.rs]: Only works for C-like enums.

**Comment by teor2345** [comment]
@gtank are you familiar with the zcash difficulty calculations?

File: zebra-chain/src/block/difficulty.rs
Comment (teor2345) [inline zebra-chain/src/block/difficulty.rs]:
```suggestion
```

File: zebra-chain/src/block/difficulty.rs
Comment (teor2345) [inline zebra-chain/src/block/difficulty.rs]:
```suggestion
```

File: zebra-chain/src/block/difficulty.rs
Comment (teor2345) [inline zebra-chain/src/block/difficulty.rs]:
```suggestion
        f.debug_tuple("Work")
        // Use hex, because expanded difficulty is in hex.
            .field(&format_args!("{:#x}", self.0))
            // Use decimal, to compare with zcashd
            .field(&format_args!("{}", self.0))
            // Use log2, to compare with zcashd
            .field(&format_args!("{:.5}", log2_work))
            .finish()
```

File: zebra-chain/src/block/difficulty.rs
Comment (teor2345) [inline zebra-chain/src/block/difficulty.rs]:
```suggestion
            // Use hex, because expanded difficulty is in hex.
```

File: zebra-chain/src/block/difficulty.rs
Comment (teor2345) [inline zebra-chain/src/block/difficulty.rs]:
```suggestion
use std::fmt;
```

File: zebra-chain/src/block/difficulty.rs
Comment (teor2345) [inline zebra-chain/src/block/difficulty.rs]:
```suggestion
            .field(&format_args!("{:.5}", (self.0 as f64).log2()))
```

File: zebra-chain/src/block/difficulty.rs
Comment (gtank) [inline zebra-chain/src/block/difficulty.rs]: Does !expanded have the same semantics as `~<arith_uint256>` in the C++ code?

File: zebra-chain/src/block/difficulty.rs
Comment (gtank) [inline zebra-chain/src/block/difficulty.rs]: This should not just handle overflows, but also map negative values (technically possible in the compact format) to zeros/None. It looks like in our code this is handled in the ExpandedDifficulty type, but `GetBlockProof` checks this again at this point.

**Review by gtank — COMMENTED** [review]
The tests for the difficulty type cover a lot of edge cases but I don't see that they address negative values, which are technically possible and should be discarded. There are two byte sequences that cover a negative value and its inverse here https://developer.bitcoin.org/reference/block_chain.html#target-nbits

File: zebra-chain/src/block/difficulty.rs
Comment (yaahc) [reply zebra-chain/src/block/difficulty.rs]: It's not specified, the behaviour of not is controlled by the `Not` trait

https://doc.rust-lang.org/std/ops/trait.Not.html

which is defined for `U256` here https://docs.rs/uint/0.8.3/src/uint/uint.rs.html#1375-1387

which then forwards to the `Not` impl for `u64` here https://doc.rust-lang.org/src/core/ops/bit.rs.html#47-52

which, extremely notably, defines the implementation of `Not` as `!self`, which just means "the compiler handles this via magic"

According to the reference however the not operator on `Integer` types is implemented as a Bitwise NOT https://doc.rust-lang.org/stable/reference/expressions/operator-expr.html#negation-operators

So it seems the answer is yes, same semantics in practice.

File: zebra-chain/src/block/difficulty.rs
Comment (teor2345) [reply zebra-chain/src/block/difficulty.rs]: If `expanded` is negative or zero, then the `if let Some(expanded) = expanded {` condition will skip this calculation, and return `None`. So I don't think there's any reason for us to duplicate those checks -  they are already performed by `to_expanded`.

That just leaves the case where a division results in a zero. In practice, I don't know if that's possible in valid blocks, due to the way the calculations are defined, and the bounds on those calculations.

If it is possible, then it looks like `GetBlockProof` will happily return zero, and main.cpp uses that result without any more checks. So we can do the same:
https://github.com/zcash/zcash/blob/d6f527a88f145fa5d233e2053eee5650775600b7/src/pow.cpp#L132

**Comment by yaahc** [comment]
Teor whats the status of this PR? Are you waiting on final approval or is there still work to be done?

**Comment by teor2345** [comment]
> The tests for the difficulty type cover a lot of edge cases but I don't see that they address negative values, which are technically possible and should be discarded. There are two byte sequences that cover a negative value and its inverse here https://developer.bitcoin.org/reference/block_chain.html#target-nbits

@yaahc I am working on including the suggested test vectors, and increasing the number of negative value tests for `Work`.

**Comment by teor2345** [comment]
> The tests for the difficulty type cover a lot of edge cases but I don't see that they address negative values, which are technically possible and should be discarded. There are two byte sequences that cover a negative value and its inverse here https://developer.bitcoin.org/reference/block_chain.html#target-nbits

There are some existing negative value tests, but some of them aren't run on `Work`:
https://github.com/ZcashFoundation/zebra/pull/838/files#diff-661319c9fdc6c7336370f7e682f8326bR101
See a6d83159ab3e610f946f3d260ca5ea97450b2ae1 for this fix.

I added all the bitcoin difficulty test vectors in 08b94600a700383d111c336daeca68258aed4a0c.

File: zebra-chain/src/block/difficulty.rs
Comment (yaahc) [inline zebra-chain/src/block/difficulty.rs]: This wont compile because of the block change so you probably dont want to apply this via github but you can use the try operator on Options


```suggestion
        let expanded = self.to_expanded()?;
```

File: zebra-chain/src/block/difficulty.rs
Comment (yaahc) [reply zebra-chain/src/block/difficulty.rs]: sounds resolved so I'm marking it as such

**Review by yaahc — DISMISSED** [review]
approved but I think you should still apply the final suggestion pre merge

File: zebra-chain/src/block/difficulty.rs
Comment (teor2345) [inline zebra-chain/src/block/difficulty.rs]:
```suggestion
        // We need to compute `2^256 / (expanded + 1)`, but we can't represent
        // 2^256, as it's too large for a u256. However, as 2^256 is at least as
        // large as `expanded + 1`, it is equal to
        // `((2^256 - expanded - 1) / (expanded + 1)) + 1`, or
        let result = (!expanded.0 / (expanded.0 + 1)) + 1;
        if result <= u128::MAX.into() {
            return Some(Work(result.as_u128()));
        }
        
        None
```

File: zebra-chain/src/block/difficulty.rs
Comment (yaahc) [inline zebra-chain/src/block/difficulty.rs]:
```suggestion

```

Comment (teor2345) [inline zebra-chain/src/block/difficulty.rs]:
```suggestion

```

**Review by yaahc — DISMISSED** [review]

File: zebra-chain/src/block/difficulty.rs
Comment (yaahc) [inline zebra-chain/src/block/difficulty.rs:235]:
```suggestion
        }

```

**Review by yaahc — APPROVED** [review]

**Pull request merged**
